### PR TITLE
[z/OS][libc++] Remove `align_val_t` dependency in small_buffer.h 

### PR DESCRIPTION
**Pull request by zibi2**
We need to guard 2 functions to avoid errors when `small_buffer.h` is included in the modules LIT tests. For example:

```
test-suite-install/include/c++/v1/__utility/small_buffer.h:69:81: error: use of undeclared identifier 'align_val_t'
# |    69 |       byte* __allocation = static_cast<byte*>(::operator new[](sizeof(_Stored), align_val_t{alignof(_Stored)}));
# |       |                                                                                 ^
```

### CODE REVIEW

**Review by abhina-sree — APPROVED** [review]
LGTM

**Review by frederick-vs-ja — COMMENTED** [review]
Currently this approach just disables dynamic allocation when aligned allocation is not available in such a case. I guess we should manually align the buffer ptrs instead.

File: libcxx/include/__utility/small_buffer.h [hunk 1]
```diff
@@ -61,6 +61,7 @@ class __small_buffer {
       return *std::launder(reinterpret_cast<_Stored**>(__buffer_));
   }
 
+#  ifdef _LIBCPP_HAS_LIBRARY_ALIGNED_ALLOCATION
```
Comment (ldionne) [inline libcxx/include/__utility/small_buffer.h:64]: As @frederick-vs-ja mentioned, we should either align the memory manually or maybe just use the unaligned version of `operator new`. As-is, the rest of the code would not even compile since `__alloc()` wouldn't be present.

I believe `__libcpp_allocate` is exactly what we want here, since it handles the fallback to aligned allocation if available.

File: libcxx/include/__utility/small_buffer.h [hunk 2]
```diff
@@ -72,11 +73,14 @@ class __small_buffer {
     }
   }
 
+#  endif // _LIBCPP_HAS_LIBRARY_ALIGNED_ALLOCATION
+#  ifdef _LIBCPP_HAS_LIBRARY_SIZED_DEALLOCATION
   template <class _Stored>
   _LIBCPP_HIDE_FROM_ABI void __dealloc() noexcept {
     if constexpr (!__fits_in_buffer<_Stored>)
       ::operator delete[](*reinterpret_cast<void**>(__buffer_), sizeof(_Stored), align_val_t{alignof(_Stored)});
```
Comment (ldionne) [inline libcxx/include/__utility/small_buffer.h:81]: Same here, I think we want to use `__libcpp_deallocate` instead, which handles the absence of sized deallocation.

**Review by ldionne — CHANGES_REQUESTED** [review]

**Comment by zibi2** [comment]
Thank you for the comments, Louis please check if I addressed your suggestions correctly.

File: libcxx/include/__utility/small_buffer.h [hunk 3]
```diff
@@ -66,16 +66,24 @@ class __small_buffer {
     if constexpr (__fits_in_buffer<_Stored>) {
       return std::launder(reinterpret_cast<_Stored*>(__buffer_));
     } else {
+#  ifdef _LIBCPP_HAS_LIBRARY_ALIGNED_ALLOCATION
       byte* __allocation = static_cast<byte*>(::operator new[](sizeof(_Stored), align_val_t{alignof(_Stored)}));
       std::construct_at(reinterpret_cast<byte**>(__buffer_), __allocation);
       return std::launder(reinterpret_cast<_Stored*>(__allocation));
+#  else
+      return static_cast<_Stored*>(std::__libcpp_allocate(_BufferSize * sizeof(byte), _LIBCPP_ALIGNOF(byte)));
+#  endif // _LIBCPP_HAS_LIBRARY_ALIGNED_ALLOCATION
```
Comment (ldionne) [inline libcxx/include/__utility/small_buffer.h:75]: I think what we want is this:

```c++
return static_cast<_Stored*>(std::__libcpp_allocate(_BufferSize * sizeof(byte), alignof(_Stored)));
```

We're always passing the right alignment, but we let `__libcpp_allocate` discard it based on `_LIBCPP_HAS_LIBRARY_ALIGNED_ALLOCATION`. So we don't need any `#if` here.

File: libcxx/include/__utility/small_buffer.h [hunk 4]
```diff
@@ -66,16 +66,24 @@ class __small_buffer {
...
   }
 
   template <class _Stored>
   _LIBCPP_HIDE_FROM_ABI void __dealloc() noexcept {
     if constexpr (!__fits_in_buffer<_Stored>)
+#  ifdef _LIBCPP_HAS_LIBRARY_SIZED_DEALLOCATION
       ::operator delete[](*reinterpret_cast<void**>(__buffer_), sizeof(_Stored), align_val_t{alignof(_Stored)});
+#  else
+      std::__libcpp_deallocate((void*)__buffer_, _BufferSize * sizeof(byte), _LIBCPP_ALIGNOF(_Stored));
+#  endif // _LIBCPP_HAS_LIBRARY_SIZED_DEALLOCATION
```
Comment (ldionne) [inline libcxx/include/__utility/small_buffer.h:86]: Similarly, here I would do this unconditionally (without `#if`):

```c++
std::__libcpp_deallocate((void*)__buffer_, _BufferSize * sizeof(byte), alignof(_Stored));
```

File: libcxx/include/__utility/small_buffer.h [hunk 5]
```diff
@@ -66,16 +66,14 @@ class __small_buffer {
     if constexpr (__fits_in_buffer<_Stored>) {
       return std::launder(reinterpret_cast<_Stored*>(__buffer_));
     } else {
-      byte* __allocation = static_cast<byte*>(::operator new[](sizeof(_Stored), align_val_t{alignof(_Stored)}));
-      std::construct_at(reinterpret_cast<byte**>(__buffer_), __allocation);
-      return std::launder(reinterpret_cast<_Stored*>(__allocation));
+      return static_cast<_Stored*>(std::__libcpp_allocate(_BufferSize * sizeof(byte), alignof(_Stored)));
```
Comment (philnik777) [inline libcxx/include/__utility/small_buffer.h:69]: This change looks wrong to me. We want to store the pointer inside the `__small_buffer` and then return the pointer to the storage. Right now we're just throwing away the pointer to the storage, since the caller isn't supposed to save it.

**Review by philnik777 — CHANGES_REQUESTED** [review]

File: libcxx/include/__utility/small_buffer.h (hunk 5 above)
Comment (ldionne) [reply libcxx/include/__utility/small_buffer.h:69]: Ah, yes, my suggestion in https://github.com/llvm/llvm-project/pull/114396#discussion_r1828007471 was incorrect (it was based on the already incorrect diff). I think this is what we want (which is a mix between my suggestion and the status quo):

```c++
byte* __allocation = static_cast<byte*>(std::__libcpp_allocate(sizeof(_Stored), alignof(_Stored)));
std::construct_at(reinterpret_cast<byte**>(__buffer_), __allocation);
return std::launder(reinterpret_cast<_Stored*>(__allocation));
```

Comment (zibi2) [reply libcxx/include/__utility/small_buffer.h]: Thank you Louis and Nikolas !!!

File: libcxx/include/__utility/small_buffer.h [hunk 6]
```diff
@@ -75,7 +75,7 @@ class __small_buffer {
   template <class _Stored>
   _LIBCPP_HIDE_FROM_ABI void __dealloc() noexcept {
     if constexpr (!__fits_in_buffer<_Stored>)
-      ::operator delete[](*reinterpret_cast<void**>(__buffer_), sizeof(_Stored), align_val_t{alignof(_Stored)});
+      std::__libcpp_deallocate((void*)__buffer_, _BufferSize * sizeof(byte), alignof(_Stored));
```
Comment (ldionne) [inline libcxx/include/__utility/small_buffer.h:78]: This is incorrect, since you're deallocating the buffer itself (which is a member variable), not the pointer contained in the buffer. The previous `reinterpret_cast` must be kept.

Comment (zibi2) [reply libcxx/include/__utility/small_buffer.h]: Thanks for caching it.

File: libcxx/include/__utility/small_buffer.h [hunk 7]
```diff
@@ -75,7 +75,7 @@ class __small_buffer {
   template <class _Stored>
   _LIBCPP_HIDE_FROM_ABI void __dealloc() noexcept {
     if constexpr (!__fits_in_buffer<_Stored>)
-      ::operator delete[](*reinterpret_cast<void**>(__buffer_), sizeof(_Stored), align_val_t{alignof(_Stored)});
+      std::__libcpp_deallocate(*reinterpret_cast<void**>(__buffer_), _BufferSize * sizeof(byte), alignof(_Stored));
```
Comment (philnik777) [inline libcxx/include/__utility/small_buffer.h:78]: Why did you change the size calculation here?

**Review by philnik777 — APPROVED** [review]

**Review by ldionne — APPROVED** [review]

**Comment by ldionne** [comment]
Let's merge once CI is green.

File: libcxx/include/__utility/small_buffer.h (hunk 7 above)
Comment (zibi2) [reply libcxx/include/__utility/small_buffer.h]: Thank you for pointing out this. The good part is that this issue was also caught by stage3 sanitizer failure.
